Bug 1850060: Make DNS queries for egress network policy async

juanluisvaladas opened the pull request:
Prior to this PR we have exactly one goroutine to query all the DNS names synchronously.
If a DNS query is slow we'll block the whole synchronization goroutine causing the IP addresses of flows for the rest of the dnsNames to be potentially outdated.

juanluisvaladas commented:
/retitle [WIP]Bug1850060: Make DNS queries for egress network policy async

juanluisvaladas commented:
/assign @abhat

abhat commented on pkg/network/common/egress_dns.go:
```diff
@@ -117,43 +121,34 @@ func (e *EgressDNS) Sync() {
-					e.Updates <- updates
-				}
-				continue
+				go e.Update(dnsName)
```
why not just run go routine for doing the dns query?

juanluisvaladas replied:
Prior to this PR we have the function `egressDNS.Sync()`, which calls `egressDNS.Update()`, which calls `DNS.Update()`. which gets blocked on `DNS.getMinIPsAndMinTTL()`.

Perhaps I'm missing something; but in order to do one goroutine per query the only two functions that can be ran in a separate goroutine are `egressDNS.Update()` or `DNS.Update()` because if we call `DNS.getMinIPsAndMinTTL()` in a separate goroutine `DNS.Update()` will still need to wait for it prior to doing the return. So instead of blocking in the function we'll block on waiting for it to write to the channel.

We could call `DNS.Update()` instead but then it would need to write to the `EgressDNS.dnsResponse` channel directly and I don't see any actual benefit. It would be harder to test and we would need to modify the tests which I managed to get without modifications.

juanluisvaladas commented:
Moving back to WIP.
After some manual testing I've identified a bottleneck:

```
W0128 11:57:36.753447   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 5.542399362s. Expected less than 100ms
W0128 11:57:47.305604   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.551092731s. Expected less than 100ms
W0128 11:57:57.884220   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.576496838s. Expected less than 100ms
W0128 11:58:08.464444   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.575839282s. Expected less than 100ms
W0128 11:58:19.042831   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.576980075s. Expected less than 100ms
W0128 11:58:29.623615   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.57840838s. Expected less than 100ms
W0128 11:58:40.200012   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.57425184s. Expected less than 100ms
W0128 11:58:50.777478   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.390045228s. Expected less than 100ms
W0128 11:59:01.355494   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.576549115s. Expected less than 100ms
W0128 11:59:11.933716   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.577143096s. Expected less than 100ms
W0128 11:59:22.512446   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.578068683s. Expected less than 100ms
W0128 11:59:33.107216   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.591565244s. Expected less than 100ms
W0128 11:59:43.696383   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.586992445s. Expected less than 100ms
W0128 11:59:54.273864   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.576080032s. Expected less than 100ms
W0128 12:00:04.851594   97626 dns.go:293] Writing to egressDNS Updates channel took too long: 10.576523652s. Expected less than 100ms
```
I don't know why yet.

juanluisvaladas commented:
I determined the previous cluster had an issue with most of the CPU stolen and everything was unreasonably slow.

I'm not able to reproduce anymore.

juanluisvaladas commented:
/assign @danwinship

abhat commented on pkg/network/common/dns.go:
```diff
@@ -24,6 +27,9 @@ type dnsValue struct {
 	nextQueryTime time.Time
+	// Used to know if DNS.Update or DNS.Add are modifying it, so that
+	// DNS.GetNextQueryTime can ignore it
+	updating bool
```
should this be atomic?

juanluisvaladas replied:
I don't think that's necessary because we're only modifying this variable when `DNS.lock` is held. So I think this is thread safe as is.

danwinship replied:
> I don't think that's necessary because we're only modifying this variable when `DNS.lock` is held. So I think this is thread safe as is.

It's not necessary because you're only modifying _or reading_ the variable when the lock is held. It wouldn't be safe to read `updating` from `GetNextQueryTime` if you weren't holding the lock there.

abhat replied:
In fact, if the only reason we hold the lock is to update the "updating" bool, we should instead make it atomic and get rid of the locks.

juanluisvaladas replied:
> In fact, if the only reason we hold the lock is to update the "updating" bool, we should instead make it atomic and get rid of the locks.

The problem is that this variable is part of a struct in a map and we cannot guarantee that the map entry won't be deleted while we update it without locking the whole map.

abhat replied:
I think that's fair. You cannot control the object's map membership by making the variable atomic.

abhat commented on pkg/network/common/dns.go:
```diff
@@ -84,14 +99,23 @@ func (d *DNS) Get(dns string) dnsValue {
+	}
+
+	defer func(start time.Time) {
+		warnSlowOperation("Add resolved egress DNS record", start)
```
shouldn't the defer func be declared before calling getIPsAndMinTTL() if you want to warn of slow operation when making the dns query?

juanluisvaladas replied:
What I wanted to log was precisely the bits that we can control and optimize (if the DNS is slow, we can't do anything about it).

Said so I agree with logging slow DNS queries is useful as well, but I'm not quite sure what's an appropriate threshold for that. I tried doing a few DNS queries to 8.8.8.8 and 1.1.1.1 and I'm getting somewhere between 20 and 80 ms most of the time. Do 300 milliseconds sound like a reasonable

abhat commented on pkg/network/common/dns.go:
```diff
@@ -101,34 +125,69 @@ func (d *DNS) Delete(dns string) {
 }
 
-func (d *DNS) Update(dnsName string) (bool, error) {
+func (d *DNS) LockForUpdate(dns string) error {
```
`s/LockForUpdate/SetUpdating`

juanluisvaladas replied:
That's definitely a better name

abhat commented on pkg/network/common/dns.go:
```diff
@@ -101,34 +125,69 @@ func (d *DNS) Delete(dns string) {
 	res.ips = ips
 	res.ttl = ttl
 	res.nextQueryTime = time.Now().Add(res.ttl)
+	res.updating = false
```
why do it twice if we are also doing it as a part of `updateNextQueryTime`?

juanluisvaladas replied:
We shouldn't do it twice for the same update. If `getIPsAndMinTTL` returns an error we call `updateNextQueryTime`, else we call `updateDNSValue`.

juanluisvaladas commented on pkg/network/common/dns.go:
```diff
@@ -70,6 +81,10 @@ func (d *DNS) Size() int {
 
 func (d *DNS) Get(dns string) dnsValue {
+	defer func(start time.Time) {
+		warnSlowOperation("Get egress DNS record", start)
```
Good idea.

juanluisvaladas commented:
/retitle Bug 1850060: Make DNS queries for egress network policy async

danwinship commented:
> _No description provided._

please fix

danwinship commented on pkg/network/common/egress_dns.go:
```diff
@@ -98,17 +102,19 @@ func (e *EgressDNS) Delete(policy networkv1.EgressNetworkPolicy) {
-func (e *EgressDNS) Update(dns string) (bool, error) {
-	e.lock.Lock()
-	defer e.lock.Unlock()
+func (e *EgressDNS) Update(dns string) {
```
if this is only used internally (which it seems like it is?) it should have a lowercase name

danwinship commented on pkg/network/common/dns.go:
```diff
@@ -347,3 +361,10 @@ func removeDuplicateIPs(ips []net.IP) []net.IP {
 	return uniqueIPs
 }
+
+func warnSlowOperation(operation string, start time.Time) {
```
dunno if you saw my response on slack the other day, but the thing you were looking for is `k8s.io/utils/trace`

juanluisvaladas replied:
Note to myself: Not  outdated TODO

danwinship commented on pkg/network/common/egress_dns.go:
```diff
@@ -133,7 +133,9 @@ func (e *EgressDNS) Sync() {
 				// TODO: Should this be non-blocking?
+				start := time.Now()
 				e.Updates <- e.getEgressDNSUpdates(response.Name)
+				warnSlowOperation("Writing to egressDNS Updates channel", start)
```
You could also rewrite this to use a non-blocking channel that's just used to signal the presence of updates, and a `GetUpdates()` method to get the actual changes. Like in #139.

juanluisvaladas replied:
Note to myself: Not  outdated TODO

juanluisvaladas replied:
Now this is done on a separate goroutine, technically it's not the same thing but it achieves the same.

danwinship reviewed (commented):
Some initial review but it would be good to have a description of exactly what is going wrong in 1850060, and how this fixes it

juanluisvaladas commented:
There is some kind of race condition in the unit test or the code itself, sometimes it passes. Trying to figure out the problem.

juanluisvaladas commented:
@danwinship I addressed all your previous concerns, added a test that forges out of order answers proving the egressDNS is handling the slow async queries correctly. Do you mind having another look at this?

abhat commented on pkg/network/common/dns.go:
```diff
@@ -84,53 +107,102 @@ func (d *DNS) Get(dns string) dnsValue {
+		return err
+	}
+
+	trace := utiltrace.New("Update resolved DNS record")
```
add the dns string name to the trace?

abhat commented on pkg/network/common/dns.go:
```diff
@@ -84,53 +107,102 @@ func (d *DNS) Get(dns string) dnsValue {
 }
 
-func (d *DNS) Update(dnsName string) (bool, error) {
+func (d *DNS) SetUpdate(dns string) error {
```
s/SetUpdate/SetUpdating

abhat commented on pkg/network/common/dns.go:
```diff
@@ -84,53 +107,102 @@ func (d *DNS) Get(dns string) dnsValue {
 
-func (d *DNS) Update(dnsName string) (bool, error) {
+func (d *DNS) SetUpdate(dns string) error {
+	trace := utiltrace.New("SetUpdate DNS record")
```
add dns name to the trace

juanluisvaladas commented:
All the comments have been addressed.

juanluisvaladas commented:
/test e2e-aws

danwinship commented on pkg/network/common/dns.go:
```diff
@@ -139,31 +139,32 @@ func (d *DNS) Update(dns string) (bool, error) {
 		// Should not happen, all operations on dnsMap are synchronized by d.lock
-		return fmt.Errorf("DNS value not found in dnsMap for domain: %q", dns)
+		klog.Errorf("DNS value not found in dnsMap for domain: %q", dns)
+	} else {
```
would be better to `return` rather than putting the rest of the function in an `else`

danwinship commented on pkg/network/common/dns.go:
```diff
@@ -264,6 +281,10 @@ func (d *DNS) queryServer(nameserver, domain string) ([]net.IP, int, error) {
 func (d *DNS) getIPsAndMinTTL(domain string) ([]net.IP, time.Duration, error) {
+	trace := utiltrace.New(fmt.Sprintf("DNS resolution for %q", domain))
+	// Probably a bit too excessive.
+	defer trace.LogIfLong(dnsQueryTraceThreshold)
```
we shouldn't be adding it if we think it's going to be excessive...

juanluisvaladas replied:
Actually what I meant is the threshold is probably too high and should be lower. I removed the comment.

danwinship commented on pkg/network/common/dns.go:
```diff
@@ -34,6 +34,15 @@ type dnsValue struct {
 	// DNS.GetNextQueryTime can ignore it
 	updating bool
 }
+type DNSInterface interface {
```
blank line before that. also, maybe use `DNS` as the name of the interface, and rename the struct? (The struct type doesn't need to be exported any more now...)

abhat replied:
+1

danwinship reviewed (changes requested):
Can you squash the first three commits? The second and third are basically fixups to the first.

abhat commented on pkg/network/common/egress_dns.go:
```diff
@@ -117,43 +131,46 @@ func (e *EgressDNS) Sync() {
+}
+
+func (e *EgressDNS) getEgressDNSUpdates(dnsName string) []EgressDNSUpdate {
+	policyUpdates := make([]EgressDNSUpdate, 0)
```
This probably needs to grab a lock since it iterates over the dnsNamesToPolicies map which Add/Delete may muck around with?

juanluisvaladas replied:
Good catch! Added the lock to it.

juanluisvaladas commented:
> Can you squash the first three commits? The second and third are basically fixups to the first.
Yep that's a good point. Also done.

abhat commented on pkg/network/common/dns.go:
```diff
@@ -218,6 +291,9 @@ func (d *DNS) queryServer(nameserver, domain string) ([]net.IP, int, error) {
 }
 
 func (d *DNS) getIPsAndMinTTL(domain string) ([]net.IP, time.Duration, error) {
```
`d.getIPsAndMinTTL` uses `d.nameservers` and `d.queryServer`. But it is not protected by a lock. It is probably OK since `d.nameservers` is not something that gets mucked around at all once the DNS object is created. And `d.queryServer` is not using any other shared resource.

juanluisvaladas replied:
> d.getIPsAndMinTTL uses d.nameservers and d.queryServer. But it is not protected by a lock

If we locked the whole function we wouldn't be able to query the DNS in parallel, making the whole patch useless.  If there were a bit unsafe to call concurrently we should lock only that bit;  the rest must be unlocked so that we can call it concurrently.

> It is probably OK since d.nameservers is not something that gets mucked around at all once the DNS object is created. And d.queryServer is not using any other shared resource.

I agree, I don't think we need to lock it at all.

juanluisvaladas commented:
/retest

juanluisvaladas commented:
/retest

abhat commented on pkg/network/common/egress_dns_test.go:
```diff
@@ -0,0 +1,123 @@
+	}
+	u0 = update[0]
+	if u0.Namespace != "fake-ns-2" {
+		t.Errorf("Expecting an update for fake-ns-1 and fake-ns-2. Got: %v", update)
```
nit: update the error statement that the update is expected only for `fake-ns-2`

abhat commented on pkg/network/common/dns.go:
```diff
@@ -84,53 +115,102 @@ func (d *DNS) Get(dns string) dnsValue {
+
 	changed := false
-	if !ipsEqual(res.ips, ips) {
+	if !d.ipsChangeGd(res.ips, dns string) {
```
typo `d.ipsChanged()`

abhat commented:
/lgtm

alexanderConstantinescu commented:
/bugzilla refresh
